### PR TITLE
In-memory Engine: implement read flow part 1

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
implement in-memory engine read flow -- part 1
```
This is the first PR implementing in-memory engine's read flow. It implements the read part of RegionCacheMemoryEngine including snapshot and iterator.
The next PR is to implement the HybridEngine part and integrate it with the raftstore.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/region_cache_memory_engine/src/engine.rs [hunk 1]
```diff
@@ -23,19 +35,86 @@ pub struct RegionMemoryEngine {
...
 impl Debug for RegionMemoryEngine {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        unimplemented!()
+        write!(f, "Region Memory Engine")
     }
 }
 
-type SnapshotList = BTreeMap<u64, u64>;
+#[derive(Default)]
+struct SnapshotList(BTreeMap<u64, u64>);
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:93]: <region_id, read_ts> may not be enough, we need ref count for this pair. Think about two snapshots with same region_id,read_ts.

File: components/region_cache_memory_engine/src/engine.rs [hunk 2]
```diff
@@ -73,22 +152,46 @@ pub struct RegionCacheMemoryEngine {
...
+        }
+
+        if read_ts <= region_meta.safe_ts {
+            // todo(SpadeA): add metrics for it
+            return None;
+        }
+
+        region_meta.snapshot_list.new_snapshot(read_ts);
+
+        Some(RegionCacheSnapshot {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:189]: it's better to have a RegionCacheSnapshot::new and then put  region_meta.snapshot_list.new_snapshot(read_ts) inside the new(). So both new_snapshot and reclaim_snapshot are called in the same place.

File: components/region_cache_memory_engine/src/engine.rs [hunk 3]
```diff
@@ -73,22 +152,46 @@ pub struct RegionCacheMemoryEngine {
...
+            .insert(region_id, RegionMemoryMeta::default());
     }
 }
 
 impl RegionCacheEngine for RegionCacheMemoryEngine {
     type Snapshot = RegionCacheSnapshot;
 
     fn snapshot(&self, region_id: u64, read_ts: u64) -> Option<Self::Snapshot> {
-        unimplemented!()
+        let mut core = self.core.lock().unwrap();
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:176]: Add // TODO here to pass seq# into the snapshot as well?

File: components/region_cache_memory_engine/src/engine.rs [hunk 4]
```diff
@@ -23,19 +35,86 @@ pub struct RegionMemoryEngine {
...
+struct SnapshotList(BTreeMap<u64, u64>);
+
+impl SnapshotList {
+    fn new_snapshot(&mut self, read_ts: u64) {
+        // snapshot with this ts may be granted before
+        let count = self.0.get(&read_ts).unwrap_or(&0) + 1;
+        self.0.insert(read_ts, count);
+    }
+
+    fn reclaim_snapshot(&mut self, read_ts: u64) {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:102]: I think reclaim is a bit misleading name, as it means re-use the snapshot. 
Maybe remove_snapshot is better.

File: components/region_cache_memory_engine/src/engine.rs [hunk 5]
```diff
@@ -23,19 +35,86 @@ pub struct RegionMemoryEngine {
...
+        let count = *self.0.get(&read_ts).unwrap();
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:103]: get_mut and update the count in-place. It's faster then re-insert with the key.

File: components/region_cache_memory_engine/src/engine.rs [hunk 6]
```diff
@@ -110,10 +213,12 @@ impl WriteBatchExt for RegionCacheMemoryEngine {
 }
 
 pub struct RegionCacheIterator {
+    cf: String,
     valid: bool,
     prefix_same_as_start: bool,
     prefix: Option<Vec<u8>>,
     iter: IterRef<Skiplist<ByteWiseComparator>, ByteWiseComparator>,
+    // The lower bound is inclusive while the upper bound is exclusive if set
     lower_bound: Option<Vec<u8>>,
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/engine.rs:222]: Is it possible to always set lower/uppper bound for this iterator, with region's start and end key if caller has no other bound.  This is the extra safety net to ensure it does not return ranges that does not belong to the region.

File: components/region_cache_memory_engine/src/engine.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:102]: Ok

File: components/region_cache_memory_engine/src/engine.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:189]: I don't understand. How can we "put region_meta.snapshot_list.new_snapshot(read_ts) inside the new()"?

File: components/region_cache_memory_engine/src/engine.rs (hunk 6 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:222]: lower/upper bound is passed through IterOptions where they are Option in it. We can unwrap here to only allow callers with lower/upper bound set to call it.

File: components/engine_traits/src/errors.rs [hunk 7]
```diff
@@ -149,6 +149,8 @@ pub enum Error {
     EntriesUnavailable,
     #[error("The entries of region is compacted")]
     EntriesCompacted,
+    #[error("The region is not available in the region cache engine")]
+    CacheNotAvailable,
```
Comment (bufferflies) [inline components/engine_traits/src/errors.rs:153]: there are some reason that causes the cache miss:
1. read_ts is too old 
2. cache is not ready 
so, Do you need to record the failed reason?

Comment (SpadeA-Tang) [reply components/engine_traits/src/errors.rs]: this error is intended to represent that the boundary of IterOptions is not set. I have adjusted the name.

File: components/region_cache_memory_engine/src/engine.rs [hunk 8]
```diff
@@ -1,19 +1,31 @@
...
 use collections::HashMap;
 use engine_traits::{
-    CfNamesExt, DbVector, IterOptions, Iterable, Iterator, Mutable, Peekable, ReadOptions,
+    CfNamesExt, DbVector, Error, IterOptions, Iterable, Iterator, Mutable, Peekable, ReadOptions,
     RegionCacheEngine, Result, Snapshot, SnapshotMiscExt, WriteBatch, WriteBatchExt, WriteOptions,
+    CF_DEFAULT, CF_LOCK, CF_WRITE,
 };
 use skiplist_rs::{ByteWiseComparator, IterRef, Skiplist};
 
+fn cf_to_id(cf: &str) -> usize {
```
Comment (bufferflies) [inline components/region_cache_memory_engine/src/engine.rs:20]: same with it https://github.com/tikv/tikv/blob/993eb2f610acfc4ee9e6c7b4f03eb63213f6d593/components/raft_log_engine/src/engine.rs#L305

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:20]: Not exactly. CF_RAFT is not included.

File: components/region_cache_memory_engine/src/engine.rs [hunk 9]
```diff
@@ -23,19 +35,87 @@ pub struct RegionMemoryEngine {
...
+                    true,
+                )),
+            ],
+        }
+    }
+}
+
+impl Default for RegionMemoryEngine {
+    fn default() -> Self {
+        RegionMemoryEngine {
```
Comment (bufferflies) [inline components/region_cache_memory_engine/src/engine.rs:64]: how about use with_capacity direct?

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:64]: fine

File: components/region_cache_memory_engine/src/engine.rs [hunk 10]
```diff
@@ -23,19 +35,87 @@ pub struct RegionMemoryEngine {
...
+impl SnapshotList {
+    fn new_snapshot(&mut self, read_ts: u64) {
+        // snapshot with this ts may be granted before
+        let count = self.0.get(&read_ts).unwrap_or(&0) + 1;
+        self.0.insert(read_ts, count);
+    }
+
+    fn remove_snapshot(&mut self, read_ts: u64) {
+        let count = self.0.get_mut(&read_ts).unwrap();
+        assert!(*count >= 1);
```
Comment (bufferflies) [inline components/region_cache_memory_engine/src/engine.rs:105]: how about add more info, such as this read_ts

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:87]: What you mean? I dont get it.

Comment (bufferflies) [reply components/region_cache_memory_engine/src/engine.rs:87]: like this  `assert!(!resp.has_error(), "{:?}", read_ts);`

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:87]: No info is exposed in new_snaphsot, so adding this info here does not help.

File: components/region_cache_memory_engine/src/engine.rs [hunk 11]
```diff
@@ -132,39 +221,84 @@ impl Iterable for RegionCacheMemoryEngine {
...
+        if self.valid && self.prefix_same_as_start {
+            // todo(SpadeA): support prefix seek
+            unimplemented!()
+        }
+        Ok(self.valid)
     }
 
     fn seek(&mut self, key: &[u8]) -> Result<bool> {
-        unimplemented!()
+        let seek_key = if key < self.lower_bound.as_slice() {
```
Comment (bufferflies) [inline components/region_cache_memory_engine/src/engine.rs:257]: Do it needs to check ` assert!(self.valid);` before this line?

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:257]: seek deos not depend on the previous operations, so it does not need it.

File: components/region_cache_memory_engine/src/engine.rs [hunk 12]
```diff
@@ -73,22 +135,47 @@ pub struct RegionCacheMemoryEngine {
...
-        unimplemented!()
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Region Cache Memory Engine")
     }
 }
 
 impl RegionCacheMemoryEngine {
-    pub fn new() -> Self {
-        RegionCacheMemoryEngine::default()
+    pub fn new_region(&self, region_id: u64) {
```
Comment (bufferflies) [inline components/region_cache_memory_engine/src/engine.rs:144]: how about rename it as "insert_region()"

File: components/region_cache_memory_engine/src/engine.rs [hunk 13]
```diff
@@ -23,19 +35,69 @@ pub struct RegionMemoryEngine {
...
+                    true,
+                )),
+            ],
+        }
+    }
+}
+
+impl Default for RegionMemoryEngine {
+    fn default() -> Self {
+        RegionMemoryEngine::with_capacity(1 << 20)
```
Comment (bufferflies) [inline components/region_cache_memory_engine/src/engine.rs:64]: using 1MB maybe more readable .

**Review by bufferflies — COMMENTED** [review]
rest LGTM

File: components/region_cache_memory_engine/src/engine.rs (hunk 12 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs:144]: new maybe semantically suitable.

File: components/region_cache_memory_engine/src/engine.rs (hunk 13 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/engine.rs]: Ok, modified.

**Review by bufferflies — APPROVED** [review]

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**